### ISSUE TITLE
Different Error Code For Cancelling Authentication Session During Sign Out

### ISSUE THREAD

**Issue by LuKenneth** [openid/AppAuth-iOS#458]
My issue revolves around signing out. When I launch the sign out session, there is a prompt to cancel it before the external agent window is presented. When this cancel button is tapped, the completion handler for the sign out session presents an error code `OIDErrorCodeUserCanceledAuthorizationFlow -3`. This is expected behavior and perfectly okay with me.

However, if the user continues and then is brought to their identity provider's page in the external agent window, there is another cancel button. When this cancel button is tapped, the completion handler for the sign out session presents the same error code
`OIDErrorCodeUserCanceledAuthorizationFlow -3`. This is an issue for me because the endpoint on the identity provider was already hit, and my identity provider's sign out implementation only requires that endpoint to be visited for sign out to be completed, so tapping the cancel button at this point does not necessarily cancel the session, but it dismisses the window.

For what it's worth, my app is a hybrid web application. I understand that this library is not directed towards hybrid web applications, but this is what I'm stuck with.

**Describe the solution you'd like**
 I would expect a different error code, something like
`OIDErrorCodeExternalUserAgentDismissed`, so that I can respond properly.

The root of the issue is that Apple's error handling for those two cancel events results in the same error code. So somehow I need to distinguish between when the user cancelled the session - did they cancel via the popup prompt asking if they would like to leave their app? Or did they cancel during the external agent window?

I've already looked into this solution, and I have a working solution that will undoubtedly require some advice/direction. I will submit a PR for this. But in summary of the solution - I check what the top view of the app is at the moment the user taps cancel, if it is on the external agent window, then we know they cancelled it at that time.

**Additional context**
Here's screenshots of the two buttons I am referencing in this issue.
![cancel_prompt](https://user-images.githubusercontent.com/11049112/67417564-156f6700-f597-11e9-85c1-4e3958c83fb3.png)
![cancel_window](https://user-images.githubusercontent.com/11049112/67417493-e6f18c00-f596-11e9-9497-a699136815b4.jpeg)

**Comment by WilliamDenniss** [comment]
I agree that distinct error codes would be ideal. However, as you noted, it doesn't appear to be feasible using documented APIs.

I added a few notes on your PR about the limitation of walking the view stack, but also the silver lining that you can use your own user agent implementations with AppAuth without forking.
